### PR TITLE
Add argument checker and tests for `utils.other_classes`, improve docstring

### PR DESCRIPTION
**Pull request by npapernot**
This follows issue #189 and should make it easier to understand the error encountered

### CODE REVIEW

**Comment by npapernot** [comment]
This also cleans up the `tests_tf` folder (some of the tests were not in the right files)

File: tests_tf/test_utils_tf.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+import numpy as np
+import unittest
+import tensorflow as tf
+
+from cleverhans import utils_tf
+
+
+def numpy_kl_with_logits(q_logits, p_logits):
+    def numpy_softmax(logits):
```
Comment (goodfeli) [inline tests_tf/test_utils_tf.py]: This version of the softmax is numerically unstable. Should subtract off the max of logits before applying the exp.

File: tests_tf/test_utils_tf.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        exp_logits = np.exp(logits)
+        return exp_logits / np.sum(exp_logits, axis=1, keepdims=True)
+
+    q = numpy_softmax(q_logits)
+    p = numpy_softmax(p_logits)
+    return (q * (np.log(q) - np.log(p))).sum(axis=1).mean()
```
Comment (goodfeli) [inline tests_tf/test_utils_tf.py]: This is not the numerically stable implementation of cross-entropy.
You need to expand
log softmax(x)_i = x_i - log sum exp(x_i)
or you can run into saturation issues.

File: tests_tf/test_utils_tf.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+from __future__ import unicode_literals
+
+import numpy as np
+import unittest
+import tensorflow as tf
+
+from cleverhans import utils_tf
+
+
+def numpy_kl_with_logits(q_logits, p_logits):
```
Comment (goodfeli) [inline tests_tf/test_utils_tf.py]: This function is already defined in tests_tf/test_utils.py, with the same mistakes. Maybe we should put one version of it in a shared location, and write some tests for it, using extreme values of logits that would break the current version without the numerical stabilization tricks.

Comment (goodfeli) [inline tests_tf/test_utils_tf.py]: It's probably a bit confusing to make q be the left argument, because definitions of the KL divergence almost always use p as the left argument. https://en.wikipedia.org/wiki/Kullback%E2%80%93Leibler_divergence

File: tests_tf/test_utils_tf.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+import numpy as np
+import unittest
+import tensorflow as tf
+
+from cleverhans import utils_tf
+
+
+def numpy_kl_with_logits(p_logits, q_logits):
+    def numpy_softmax(logits):
+        logits = (logits.transpose() - np.max(logits, axis=1)).transpose()
```
Comment (goodfeli) [inline tests_tf/test_utils_tf.py]: instead of two transposes, it's better to do keepdims=True in the max

Comment (npapernot) [inline tests_tf/test_utils_tf.py]: Thanks :)
